src/index.js

```diff
@@ -1,9 +1,151 @@
 'use strict';
 
+const Ajv = require('ajv');
+const get = require('lodash.get');
+const set = require('lodash.set');
+
+const EXCLUDE_PATHS = new Set([
+	'serverless', 'serviceObject', 'pluginsData', 'package',
+	'configValidationMode', 'disabledDeprecations', 'serviceFilename', 'app',
+	'tenant', 'org', 'initialServerlessConfig'
+]);
+
+const SCHEMA = {
+	type: 'object',
+	properties: {
+		iterator: {
+			oneOf: [
+				{
+					type: ['array']
+				},
+				{
+					type: 'object',
+					oneOf: [
+						{
+							properties: {
+								$env: {
+									type: 'string'
+								}
+							},
+							required: ['$env'],
+							additionalProperties: false
+						},
+						{
+							not: {
+								required: ['$env']
+							},
+							additionalProperties: true,
+						}
+					]
+				}
+			]
+		},
+		template: {
+			type: ['array', 'object']
+		}
+	},
+	required: ['iterator', 'template'],
+	additionalProperties: false
+};
+
 class ForEachPlugin {
-	constructor(serverless, options) {
+	constructor(serverless) {
 		this.serverless = serverless;
-		this.options = options;
+
+		this.ajv = new Ajv({ allowUnionTypes: true });
+		this.validator = this.ajv.compile(SCHEMA);
+
+		this.hooks = {
+			'after:package:initialize': this.replace.bind(this),
+		};
+
+		if (this.serverless.configSchemaHandler && this.serverless.configSchemaHandler.defineFunctionEvent) {
+			this.serverless.configSchemaHandler.defineFunctionEvent('aws', '$forEach', SCHEMA);
+		}
+	}
+
+	interpolate(obj, key, value) {
+		const stringified = JSON.stringify(obj);
+		return JSON.parse(stringified
+			.replace('$forEach.key', key)
+			.replace('$forEach.value', value)
+		);
+	}
+
+	findAndReplace(obj, path) {
+		if (Array.isArray(obj)) {
+			obj.forEach((val, idx) => this.findAndReplace(val, `${path}[${idx}]`));
+		} else if (typeof obj === 'object' && obj !== null) {
+			for (const key in obj) {
+				if (key === '$forEach') {
+					this.log(`Found a match in ${path}`);
+
+					this.validate(obj[key], `${path}/${key}`);
+
+					const { iterator: rawIterator, template } = obj[key];
+
+					const iterator = {};
+					if (rawIterator.$env) {
+						Object.entries(process.env).forEach(([name, value]) => {
+							if (name.match(rawIterator.$env)) {
+								iterator[name] = value;
+							}
+						});
+					} else if (Array.isArray(rawIterator)) {
+						rawIterator.forEach((value, idx) => iterator[idx] = value);
+					} else {
+						Object.assign(iterator, rawIterator);
+					}
+
+					const interpolated = Object.entries(iterator).reduce((acc, [key, value]) => {
+						const entry = this.interpolate(template, key, value);
+
+						return Array.isArray(template)
+							? acc.concat(entry)
+							: { ...acc, ...entry };
```

Comment:
Should we fail if duplicate keys are found.

Reply:
Not sure about this. I can think of use cases when someone may want to override a key value, though it's rare.

Reply:
Actually I think it's impossible to have duplicate entries here. L96 is where we should do the duplicate check, since that's really the only place where any duplicates are overridden.

Reply:
you can easily have duplicates if you do something like
```yaml
$forEach:
  iterator: [1, 1, 1]
  template:
    $forEach.value: $forEach.value
```

Reply:
ah, seems like a bad configuration 🤷. Doesn't really matter to me. It just seems like it would help debugging.